### PR TITLE
add xiaomi door window sensor 2

### PR DESCRIPTION
**Pull request by Hmm466**
# What does this implement/fix?
add xiaomi door window sensor 2
Quick description and explanation of changes

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2254

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
binary_sensor:
  - platform: xiaomi_mccgq02hl
    name: "xiaomi_mccgq02hl"
    mac_address: "xxxxxx"
    bindkey: "xxxxxxx"
    on_lock:
      name: "xiaomi_mccgq02hl"
    light:
      name: "xiaomi_mccgq02hl light"

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/xiaomi_ble/xiaomi_ble.cpp [hunk 1]
```diff
@@ -223,6 +225,9 @@ optional<XiaomiParseResult> parse_xiaomi_header(const esp32_ble_tracker::Service
     result.name = "RTCGQ02LM";
     if (raw.size() == 19)
       result.raw_offset -= 6;
+  } else if (device_uuid == 0x098B) {  // Xiaomi Mi Door window sensor 2
+    result.type = XiaomiParseResult::TYPE_MCCGQ02HL;
+    result.name = "TYPE_MCCGQ02HL";
```
Comment (jesserockz) [inline esphome/components/xiaomi_ble/xiaomi_ble.cpp]:
```suggestion
    result.name = "MCCGQ02HL";
```

File: esphome/components/xiaomi_mccgq02hl/binary_sensor.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import sensor, binary_sensor, esp32_ble_tracker
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/binary_sensor.py]:
```suggestion
from esphome.components import binary_sensor, esp32_ble_tracker
```

File: esphome/components/xiaomi_mccgq02hl/binary_sensor.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    CONF_MAC_ADDRESS,
+    CONF_BINDKEY,
+    CONF_LIGHT,
+    DEVICE_CLASS_LIGHT,
+    DEVICE_CLASS_DOOR,
+    CONF_ON_LOCK,
+)
+
+DEPENDENCIES = ["esp32_ble_tracker"]
+AUTO_LOAD = ["xiaomi_ble", "sensor"]
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/binary_sensor.py:15]:
```suggestion
AUTO_LOAD = ["xiaomi_ble"]
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by Hmm466 — COMMENTED** [review]
I checked the new changes and there seems to be no problem

**Comment by jesserockz** [comment]
Please add to a test file and create a docs PR

**Comment by Hmm466** [comment]
https://github.com/esphome/esphome/pull/3725#issuecomment-1226676272
> Please add to a test file and create a docs PR

This is my first time working on a public PR submission and overall I feel a little overwhelmed, but I'm also trying to learn and refine my process.

**Comment by Hmm466** [comment]
> Please add to a test file and create a docs PR

I have added documentation and tested it locally, new devices can be detected and status can be obtained on other platforms

File: esphome/components/xiaomi_mccgq02hl/binary_sensor.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    CONF_BINDKEY,
+    CONF_LIGHT,
+    DEVICE_CLASS_LIGHT,
+    DEVICE_CLASS_DOOR,
+    CONF_ON_LOCK,
+)
+
+CODEOWNERS = ["@JordyH66"]
+DEPENDENCIES = ["esp32_ble_tracker"]
+AUTO_LOAD = ["xiaomi_ble",'sensor']
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/binary_sensor.py]:
```suggestion
AUTO_LOAD = ["xiaomi_ble", "sensor"]
```

File: esphome/components/xiaomi_ble/xiaomi_ble.h [hunk 5]
```diff
@@ -41,6 +42,7 @@ struct XiaomiParseResult {
   optional<bool> is_active;
   optional<bool> has_motion;
   optional<bool> is_light;
+  optional<bool> is_lock;
```
Comment (jesserockz) [inline esphome/components/xiaomi_ble/xiaomi_ble.h:45]: `open` is probably better than `lock` as this does not tell you if something is locked, but open or closed.
```suggestion
  optional<bool> is_open;
```

File: esphome/components/xiaomi_mccgq02hl/binary_sensor.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+CONFIG_SCHEMA = cv.All(
+    binary_sensor.binary_sensor_schema(XiaomiMCCGQ02HL, device_class=DEVICE_CLASS_DOOR)
+    .extend(
+        {
+            cv.Required(CONF_MAC_ADDRESS): cv.mac_address,
+            cv.Required(CONF_BINDKEY): cv.bind_key,
+            cv.Optional(CONF_ON_LOCK, default=True): binary_sensor.binary_sensor_schema(
+                device_class=DEVICE_CLASS_DOOR
+            ),
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/binary_sensor.py]:
```suggestion
```

File: esphome/components/xiaomi_mccgq02hl/binary_sensor.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    var = await binary_sensor.new_binary_sensor(config)
+    await cg.register_component(var, config)
+    await esp32_ble_tracker.register_ble_device(var, config)
+
+    cg.add(var.set_address(config[CONF_MAC_ADDRESS].as_hex))
+    cg.add(var.set_bindkey(config[CONF_BINDKEY]))
+
+    if CONF_ON_LOCK in config:
+        sens = await binary_sensor.new_binary_sensor(config[CONF_ON_LOCK])
+        cg.add(var.set_lock(sens))
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/binary_sensor.py]:
```suggestion
```

File: esphome/components/xiaomi_mccgq02hl/xiaomi_mccgq02hl.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace esphome {
+namespace xiaomi_mccgq02hl {
+
+static const char *const TAG = "xiaomi_mccgq02hl";
+
+void XiaomiMCCGQ02HL::dump_config() {
+  ESP_LOGCONFIG(TAG, "Xiaomi MCCGQ02HL");
+  LOG_BINARY_SENSOR("  ", "Lock", this->is_lock_);
+  LOG_BINARY_SENSOR("  ", "Light", this->is_light_);
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/xiaomi_mccgq02hl.cpp:14]:
```suggestion
  LOG_BINARY_SENSOR("", "Xiaomi MCCGQ02HL", this);
  LOG_BINARY_SENSOR("  ", "Light", this->light_);
```

File: esphome/components/xiaomi_mccgq02hl/xiaomi_mccgq02hl.cpp [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    if (!(xiaomi_ble::parse_xiaomi_message(service_data.data, *res))) {
+      continue;
+    }
+    if (!(xiaomi_ble::report_xiaomi_results(res, device.address_str()))) {
+      continue;
+    }
+    if (res->is_light.has_value() && this->is_light_ != nullptr)
+      this->is_light_->publish_state(*res->is_light);
+    if (res->is_lock.has_value() && this->is_lock_ != nullptr)
+      this->is_lock_->publish_state(!*res->is_lock);
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/xiaomi_mccgq02hl.cpp:46]:
```suggestion
    if (res->is_light.has_value() && this->light_ != nullptr)
      this->light_->publish_state(*res->is_light);
    if (res->is_open.has_value())
      this->publish_state(*res->is_open);
```

File: esphome/components/xiaomi_mccgq02hl/xiaomi_mccgq02hl.h [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ public:
+  void set_address(uint64_t address) { address_ = address; }
+  void set_bindkey(const std::string &bindkey);
+
+  bool parse_device(const esp32_ble_tracker::ESPBTDevice &device) override;
+
+  void dump_config() override;
+  float get_setup_priority() const override { return setup_priority::DATA; }
+  void set_light(binary_sensor::BinarySensor *light) { is_light_ = light; }
+  void set_lock(binary_sensor::BinarySensor *lock) { is_lock_ = lock; }
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/xiaomi_mccgq02hl.h:26]:
```suggestion
  void set_light(binary_sensor::BinarySensor *light) { this->light_ = light; }
```

File: esphome/components/xiaomi_mccgq02hl/xiaomi_mccgq02hl.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+ protected:
+  uint64_t address_;
+  uint8_t bindkey_[16];
+  binary_sensor::BinarySensor *is_light_{nullptr};
+  binary_sensor::BinarySensor *is_lock_{nullptr};
```
Comment (jesserockz) [inline esphome/components/xiaomi_mccgq02hl/xiaomi_mccgq02hl.h:32]:
```suggestion
  binary_sensor::BinarySensor *light_{nullptr};
```

File: tests/test2.yaml [hunk 12]
```diff
@@ -446,6 +446,14 @@ binary_sensor:
       name: 'Mi Motion Sensor 2 Light'
     button:
       name: 'Mi Motion Sensor 2 Button'
+  - platform: xiaomi_mccgq02hl
+    name: 'MCCGQ02HL Activation State'
+    mac_address: '12:34:56:12:34:56'
+    bindkey: '48403ebe2d385db8d0c187f81e62cb64'
+    on_lock:
+      name: 'Mi Door Window 2 Door'
```
Comment (jesserockz) [inline tests/test2.yaml]:
```suggestion
    name: 'Mi Door Window 2 Door'
    mac_address: '12:34:56:12:34:56'
    bindkey: '48403ebe2d385db8d0c187f81e62cb64'
```

File: esphome/components/xiaomi_ble/xiaomi_ble.cpp [hunk 13]
```diff
@@ -80,6 +80,8 @@ bool parse_xiaomi_value(uint16_t value_type, const uint8_t *data, uint8_t value_
     result.has_motion = !idle_time;
   } else if ((value_type == 0x1018) && (value_length == 1)) {
     result.is_light = data[0];
+  } else if ((value_type == 0x1019) && (value_length == 1)) {
+    result.is_lock = data[0];
```
Comment (jesserockz) [inline esphome/components/xiaomi_ble/xiaomi_ble.cpp:84]:
```suggestion
    result.is_open = data[0];
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by jesserockz** [comment]
For my review above, you dont need a second child binary sensor as the main code **is** a binary sensor and just needs to publish to itself.

Also, `open` suits this id better than `lock`

**Comment by Hmm466** [comment]
Sorry, it was my negligence. I was also looking for a very close word to express it at the time, but I was afraid that my changes would affect the project, so I used ON_LOCK. So when I feel that I have become more and more aware of the project when I created a new config [CONFIG_DOOR] to replace

**Comment by Hmm466** [comment]
> Sorry, it was my negligence. I was also looking for a very close word to express it at the time, but I was afraid that my changes would affect the project, so I used ON_LOCK. So when I feel that I have become more and more aware of the project when I created a new config [CONFIG_DOOR] to replace

I think it's OK to reject my submission, OPEN seems to be more suitable for the real role of this sensor than DOOR

**Comment by jesserockz** [comment]
> Sorry, it was my negligence. I was also looking for a very close word to express it at the time, but I was afraid that my changes would affect the project, so I used ON_LOCK. So when I feel that I have become more and more aware of the project when I created a new config [CONFIG_DOOR] to replace

You miss the point I was making, The main component **IS** already the main door binary sensor, you do not need a sub-sensor named door

**Pull request closed without merging**